### PR TITLE
Upgrade dependencies and align yarn.lock

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes

Upgrade dependencies in package.json
Upgrade EsLint
Upgrade Typescript
Locked version for `fp-ts` and `io-ts`

#### How Has This Been Tested?

Unit test and lint

#### Types of changes

- [X] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: src/requests.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ export type BasicResponseHeaderKey =
   | "Transfer-Encoding";
 
 export type ResponseHeaders<H extends string = never> = {
-  [key in BasicResponseHeaderKey | H]?: string;
+  readonly [key in BasicResponseHeaderKey | H]?: string;
```
Comment (arcogabbo) [inline src/requests.ts:82]: is this wanted?

Comment (BurnedMarshal) [reply src/requests.ts]: Fixed in dd3cfe3

File: src/types.ts [hunk 2]
```diff
@@ -108,12 +108,12 @@ export const readonlyNonEmptySetType = <E>(
...
   name: string
 ): t.Type<ReadOnlyNonEmptySet<E>, t.mixed> =>
-  t.refinement(readonlySetType(o, name), e => e.size > 0, name);
+  t.refinement(readonlySetType(o, name), (e) => e.size > 0, name);
 
 /**
  * Returns a new type that has only the F fields of type T.
  */
-export type LimitedFields<T, F extends keyof T> = { [P in F]: T[P] };
+export type LimitedFields<T, F extends keyof T> = { readonly [P in F]: T[P] };
```
Comment (arcogabbo) [inline src/types.ts:116]: maybe this is another unwanted readonly

File: src/types.ts [hunk 3]
```diff
@@ -222,13 +222,13 @@ export const withDefault = <T extends t.Any>(
  */
 export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
 
-export type ReplaceProp1<T, P1 extends keyof T, A> = { [K in P1]: A } &
-  Pick<T, Exclude<keyof T, P1>>;
+export type ReplaceProp1<T, P1 extends keyof T, A> = {
+  readonly [K in P1]: A;
```
Comment (arcogabbo) [inline src/types.ts:226]: also here

Comment (BurnedMarshal) [reply src/types.ts]: fixed in f98fb5d

File: src/types.ts [hunk 4]
```diff
@@ -222,13 +222,13 @@ export const withDefault = <T extends t.Any>(
...
+} & Pick<T, Exclude<keyof T, P1>>;
 
 export type ReplaceProp2<T, P1 extends keyof T, P2 extends keyof T[P1], A> = {
-  [K in P1]: ReplaceProp1<T[K], P2, A>;
-} &
-  Pick<T, Exclude<keyof T, P1>>;
+  readonly [K in P1]: ReplaceProp1<T[K], P2, A>;
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/types.ts]: fixed in f98fb5d

**Review by arcogabbo — APPROVED** [review]

**Review by gquadrati — APPROVED** [review]
LGTM

**Pull request merged**